File: Euclid/Game/Item/Item.cs
using Euclid.Game.Entity.Humanoid;
using Euclid.Network.Streams.Util;
using Euclid.Storage.Database.Data;
using System;
using System.Collections.Generic;

namespace Euclid.Game
{
    public class Item : IEntity
    {
        #region Properties

        public ItemData Data { get; }
        public virtual ItemDefinition Definition { get { return ItemManager.Instance.GetDefinition(Data.DefinitionId); } set { Definition = value; } }
        public Room Room => RoomManager.Instance.GetRoom(Data.RoomId);
        public RoomTile CurrentTile => Position != null ? (Position.GetTile(Room) ?? null) : null;
        public Position Position { get; set; }
        public string PaddedId => Convert.ToString(Data.Id).PadLeft(Sprite.Length, '0');
        public string Sprite => Definition.Data.Sprite.Contains("*") ? Definition.Data.Sprite.Split("*")[0] : Definition.Data.Sprite;
        #endregion

        #region Constructors

        public Item(ItemData data)
        {
            Data = data;
            Position = new Position(data.X, data.Y, data.Z, data.Rotation, data.Rotation);
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Serialise item for rooms
        /// </summary>
        public void Serialise(List<object> data)
        {
            if (Definition.Data.IsFloorItem)
            {
                data.Add(new ArgumentEntry(this.PaddedId));
                data.Add(new DelimeterEntry(this.Data.Id, string.Empty));
                data.Add(new DelimeterEntry(this.Sprite, ","));
                data.Add(new DelimeterEntry(this.Position.X, " "));
                data.Add(new DelimeterEntry(this.Position.Y, " "));
                data.Add(new DelimeterEntry(this.Definition.Data.Length, " "));
                data.Add(new DelimeterEntry(this.Definition.Data.Width, " "));
                data.Add(new DelimeterEntry(this.Position.Rotation, " "));
                data.Add(new DelimeterEntry(this.Position.Z, " "));
                data.Add(new DelimeterEntry(this.Definition.Data.Colour, " "));
                data.Add(new DelimeterEntry(this.Definition.Data.Name, "/"));
                data.Add(new DelimeterEntry(this.Definition.Data.Description, "/"));
                data.Add(new DelimeterEntry(this.Definition.Data.DataClass, "/"));
                data.Add(new DelimeterEntry(this.Data.CustomData, "/"));
            }
            else
            {
                // 6;poster;Alex;leftwall 7.1250,1.3333,7125[13]56\
                data.Add(new ArgumentEntry(this.Data.Id));
                data.Add(new DelimeterEntry(this.Sprite, ";"));
                data.Add(new DelimeterEntry(null, ";"));
                data.Add(new DelimeterEntry(this.Data.WallPosition, ";"));
                data.Add(new ArgumentEntry(this.Data.CustomData));
                data.Add(@"\");
            }
        }

        public void UpdateEntities(Position position = null)
        {
            List<Humanoid> entities = new List<Humanoid>();

            foreach (Position affectedPositon in AffectedTile.GetAffectedTiles(this))
            {
                var tile = affectedPositon.GetTile(Room);

                if (tile == null)
                    continue;

                entities.AddRange(tile.Entities.Values);
            }

            if (position != null)
            {
                foreach (Position affectedPositon in AffectedTile.GetAffectedTiles(this, position.X, position.Y, position.Rotation))
                {
                    var tile = affectedPositon.GetTile(Room);

                    if (tile == null)
                        continue;

                    entities.AddRange(tile.Entities.Values);
                }
            }

            foreach (var entity in entities)
                entity.RoomEntity.InteractItem();
        }

        /// <summary>
        /// Get whether the item is walkable
        /// </summary>
        public bool IsWalkable(Position position)
        {
            if (Definition.Data.IsWalkable)
                return true;

            if (Definition.Data.IsChair)
                return true;

            if (Definition.Data.IsBed)
                return true;

            return false;
        }

        /// <summary>
        /// Get total height for furni, which is floor z axis, plus height of furni
        /// </summary>
        public double Height => Definition.Data.Height + Position.Z + 0.001;

        /// <summary>
        /// Override the position from the data variable, used to save to the database.
        /// </summary>
        public void ApplyPosition()
        {
            Position = new Position(Data.X, Data.Y, Data.Z, Data.Rotation, Data.Rotation);
        }

        /// <summary>
        ///  Check if the move is valid before moving an item.
        /// </summary>
        public bool IsValidMove(Item item, Room room, int x, int y, int rotation)
        {
            RoomTile tile = new Position(x, y).GetTile(room);

            if (tile == null || room == null)
                return false;

            bool isRotation = (item.Position.Rotation != rotation) && (new Position(x, y) == item.Position); ;

            if (isRotation)
            {
                if (item.Definition.Data.Length <= 1 && item.Definition.Data.Width <= 1)
                {
                    return true;
                }
            }

            foreach (Position position in AffectedTile.GetAffectedTiles(this, x, y, rotation))
            {
                tile = position.GetTile(room);

                if (tile == null || !room.Model.IsTile(position))
                    return false;

                if (room.Model.TileStates[position.X, position.Y] == TileState.CLOSED)
                    return false;

                if (!isRotation/* && item.Definition.InteractorType != InteractorType.BED && item.Definition.InteractorType != InteractorType.CHAIR*/)
                {
                    if (tile.Entities.Count > 0)
                        return false;
                }

                Item highestItem = tile.HighestItem;

                if (highestItem != null && highestItem.Data.Id != item.Data.Id)
                {
                    if (!CanItemsMerge(item, highestItem, new Position(x, y)))
                        return false;

                }

                foreach (Item tileItem in tile.Furniture.Values)
                {
                    if (tileItem.Data.Id == item.Data.Id)
                        continue;

                    if (!CanItemsMerge(item, tileItem, new Position(x, y)))
                        return false;
                }
            }


            return true;
        }

        /// <summary>
        /// Get if placing an item on top of another item is allowed.
        /// </summary>
        private bool CanItemsMerge(Item item, Item tileItem, Position targetTile)
        {
            if (!tileItem.Definition.Data.IsStackable)
                return false;

            if (tileItem.Definition.Data.IsChair)
                return false;

            if (tileItem.Definition.Data.IsBed)
                return false;

            return true;
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/game/item/Item.java
package euclid.game.item;

import euclid.game.entity.IEntity;
import euclid.game.pathfinder.AffectedTile;
import euclid.game.pathfinder.Position;
import euclid.game.room.Room;
import euclid.game.room.RoomManager;
import euclid.game.room.RoomTile;
import euclid.game.room.TileState;
import euclid.game.entity.Humanoid;
import euclid.network.streams.util.ArgumentEntry;
import euclid.network.streams.util.DelimeterEntry;
import euclid.storage.database.data.ItemData;

import java.util.ArrayList;
import java.util.List;

public class Item implements IEntity {

    private final ItemData data;
    private ItemDefinition definition;
    private Position position;

    public Item(ItemData data) {
        this.data = data;
        this.position = new Position(data.getX(), data.getY(), data.getZ(), data.getRotation(), data.getRotation());
    }

    public ItemData getData() {
        return data;
    }

    public ItemDefinition getDefinition() {
        if (definition != null)
            return definition;
        return ItemManager.getInstance().getDefinition(data.getDefinitionId());
    }

    public void setDefinition(ItemDefinition definition) {
        this.definition = definition;
    }

    public Room getRoom() {
        return RoomManager.getInstance().getRoom(data.getRoomId());
    }

    public RoomTile getCurrentTile() {
        return position != null ? (position.getTile(getRoom()) != null ? position.getTile(getRoom()) : null) : null;
    }

    public Position getPosition() {
        return position;
    }

    public void setPosition(Position position) {
        this.position = position;
    }

    public String getPaddedId() {
        String id = String.valueOf(data.getId());
        int spriteLen = getSprite().length();
        if (id.length() < spriteLen) {
            id = "0".repeat(spriteLen - id.length()) + id;
        }
        return id;
    }

    public String getSprite() {
        String sprite = getDefinition().getData().getSprite();
        return sprite.contains("*") ? sprite.split("\\*")[0] : sprite;
    }

    public void serialise(List<Object> data) {
        if (getDefinition().getData().isFloorItem()) {
            data.add(new ArgumentEntry(getPaddedId()));
            data.add(new DelimeterEntry(this.data.getId(), ""));
            data.add(new DelimeterEntry(getSprite(), ","));
            data.add(new DelimeterEntry(position.getX(), " "));
            data.add(new DelimeterEntry(position.getY(), " "));
            data.add(new DelimeterEntry(getDefinition().getData().getLength(), " "));
            data.add(new DelimeterEntry(getDefinition().getData().getWidth(), " "));
            data.add(new DelimeterEntry(position.getRotation(), " "));
            data.add(new DelimeterEntry(position.getZ(), " "));
            data.add(new DelimeterEntry(getDefinition().getData().getColour(), " "));
            data.add(new DelimeterEntry(getDefinition().getData().getName(), "/"));
            data.add(new DelimeterEntry(getDefinition().getData().getDescription(), "/"));
            data.add(new DelimeterEntry(getDefinition().getData().getDataClass(), "/"));
            data.add(new DelimeterEntry(this.data.getCustomData(), "/"));
        } else {
            data.add(new ArgumentEntry(this.data.getId()));
            data.add(new DelimeterEntry(getSprite(), ";"));
            data.add(new DelimeterEntry(null, ";"));
            data.add(new DelimeterEntry(this.data.getWallPosition(), ";"));
            data.add(new ArgumentEntry(this.data.getCustomData()));
            data.add("\\");
        }
    }

    public void updateEntities(Position position) {
        List<Humanoid> entities = new ArrayList<>();

        for (Position affectedPosition : AffectedTile.getAffectedTiles(this)) {
            RoomTile tile = affectedPosition.getTile(getRoom());
            if (tile == null)
                continue;
            entities.addAll(tile.getEntities().values());
        }

        if (position != null) {
            for (Position affectedPosition : AffectedTile.getAffectedTiles(this, position.getX(), position.getY(), position.getRotation())) {
                RoomTile tile = affectedPosition.getTile(getRoom());
                if (tile == null)
                    continue;
                entities.addAll(tile.getEntities().values());
            }
        }

        for (Humanoid entity : entities)
            entity.getRoomEntity().interactItem();
    }

    public void updateEntities() {
        updateEntities(null);
    }

    public boolean isWalkable(Position position) {
        if (getDefinition().getData().isWalkable())
            return true;
        if (getDefinition().getData().isChair())
            return true;
        if (getDefinition().getData().isBed())
            return true;
        return false;
    }

    public double getHeight() {
        return getDefinition().getData().getHeight() + position.getZ() + 0.001;
    }

    public void applyPosition() {
        position = new Position(data.getX(), data.getY(), data.getZ(), data.getRotation(), data.getRotation());
    }

    public boolean isValidMove(Item item, Room room, int x, int y, int rotation) {
        RoomTile tile = new Position(x, y).getTile(room);

        if (tile == null || room == null)
            return false;

        boolean isRotation = (item.getPosition().getRotation() != rotation) && (new Position(x, y).equals(item.getPosition()));

        if (isRotation) {
            if (item.getDefinition().getData().getLength() <= 1 && item.getDefinition().getData().getWidth() <= 1) {
                return true;
            }
        }

        for (Position pos : AffectedTile.getAffectedTiles(this, x, y, rotation)) {
            tile = pos.getTile(room);

            if (tile == null || !room.getModel().isTile(pos))
                return false;

            if (room.getModel().getTileStates()[pos.getX()][pos.getY()] == TileState.CLOSED)
                return false;

            if (!isRotation) {
                if (tile.getEntities().size() > 0)
                    return false;
            }

            Item highestItem = tile.getHighestItem();

            if (highestItem != null && highestItem.getData().getId() != item.getData().getId()) {
                if (!canItemsMerge(item, highestItem, new Position(x, y)))
                    return false;
            }

            for (Item tileItem : tile.getFurniture().values()) {
                if (tileItem.getData().getId() == item.getData().getId())
                    continue;
                if (!canItemsMerge(item, tileItem, new Position(x, y)))
                    return false;
            }
        }

        return true;
    }

    private boolean canItemsMerge(Item item, Item tileItem, Position targetTile) {
        if (!tileItem.getDefinition().getData().isStackable())
            return false;
        if (tileItem.getDefinition().getData().isChair())
            return false;
        if (tileItem.getDefinition().getData().isBed())
            return false;
        return true;
    }
}
